# UT1. Ejercicios A2. POO en Python: Clases y objetos

Realizar los siguientes ejercicios en Python, cada uno en su propia celda.
No es necesario crear 2 celdas para un ejercio, cada ejercicio se resuelve en una celda.

### Ejercicio 1

Escribe una clase llamada Rectangulo que contenga:
- Dos atributos: base y altura.
- Un método que calcule el área del rectángulo (base * altura).
- Un método que calcule el perímetro del rectángulo (2 * base + 2 * altura).

Crea un objeto de la clase Rectangulo y comprueba que los métodos funcionan correctamente.

In [3]:
class Rectangulo:
    def __init__(self, base, altura):
        self.base = base
        self.altura = altura

    def Area(self):
        return (self.base * self.altura)
    
    def Perimetro(self):
        return (self.base * 2) + (self.altura * 2)
    
rectangulo1 = Rectangulo(23, 37)

print(rectangulo1.Area())
print(rectangulo1.Perimetro())

851
120


### Ejercicio 2

Escribe una clase llamada Persona que contenga:
- Un atributo llamado nombre.
- Un método que imprima por pantalla el mensaje "Hola, me llamo {nombre}".
- Un método que imprima por pantalla el mensaje "Adiós, me voy de vacaciones".

In [4]:
class Persona:
    def __init__(self, nombre):
        self.nombre = nombre
    
    def saludo(self):
        print(f"Hola, me llamo {self.nombre}.")

    def adios(self):
        print("¡¡¡Me voy de vacaciones!!!")

victor = Persona("Víctor")
victor.saludo()
victor.adios()

Hola, me llamo Víctor.
¡¡¡Me voy de vacaciones!!!


### Ejercicio 3

Dado la siguiente clase:

```python	
class Vehicle:
    def __init__(self, name, max_speed, mileage):
        self.name = name
        self.max_speed = max_speed
        self.mileage = mileage
```
Crea una clase Bus que herede de la clase Vehicle y que contenga:
- Un atributo llamado _capacidad_.
- Un método _info_ que imprima por pantalla el mensaje "Bus con capacidad para {_capacidad_} pasajeros".

In [15]:
class Vehicle:
    def __init__(self, name, max_speed, mileage):
        self.name = name
        self.max_speed = max_speed
        self.mileage = mileage

class Bus(Vehicle):
    def __init__(self, name, max_speed, mileage, capacidad):
        super().__init__(name, max_speed, mileage)
        self.capacidad = capacidad
    
    def info(self):
        print(f"Bus con capacidad para  {self.capacidad} pasajeros")

bus1 = Bus("Autobus Interprovincial", 100, 26, 50)
bus1.info()

bus2 = Bus("Autobus Urbano", 50, 36, 60)
bus2.info()

Bus con capacidad para  50 pasajeros
Bus con capacidad para  60 pasajeros


### Ejercicio 4

Basado en el ejercicio anterior, agrega a la clase Bus un constructor que reciba como parámetros los atributos de la clase Vehicle y el atributo capacidad.<br>
El atributo capacidad debe tener un valor por defecto de 50.

In [16]:
class Vehicle:
    def __init__(self, name, max_speed, mileage):
        self.name = name
        self.max_speed = max_speed
        self.mileage = mileage

class Bus(Vehicle):
    def __init__(self, name, max_speed, mileage, capacidad = 50):
        super().__init__(name, max_speed, mileage)
        self.capacidad = capacidad
    
    def info(self):
        print(f"Bus con capacidad para  {self.capacidad} pasajeros")

bus1 = Bus("Autobus Interprovincial", 100, 26)
bus1.info()

bus2 = Bus("Autobus Urbano", 50, 36, capacidad = 60)
bus2.info()

Bus con capacidad para  50 pasajeros
Bus con capacidad para  60 pasajeros


### Ejercicio 5

Basado en el ejercicio 4, agrega un atributo de clase, a la clase _Vehicle_ llamado _colorBase_ con un valor por defecto de "blanco".<br>
Crear un objeto de la clase Bus y comprobar que el atributo _colorBase_ es accesible desde el objeto.<br>
Modifica el método _info_ para incluir el color base del vehículo.

In [4]:
class Vehicle:
    def __init__(self, name, max_speed, mileage, colorBase = "Blanco"):
        self.name = name
        self.max_speed = max_speed
        self.mileage = mileage
        self.colorBase = colorBase

class Bus(Vehicle):
    def __init__(self, name, max_speed, mileage, colorBase = "Blanco", capacidad = 50):
        Vehicle.__init__(self, name, max_speed, mileage, colorBase)
        self.capacidad = capacidad
    
    def info(self):
        print(f"Bus con capacidad para  {self.capacidad} pasajeros y de color {self.colorBase}")

bus1 = Bus("Autobus Interprovincial", 100, 26)
bus1.info()

bus2 = Bus("Autobus Urbano", 50, 36, capacidad = 60)
bus2.info()

Bus con capacidad para  50 pasajeros y de color Blanco
Bus con capacidad para  60 pasajeros y de color Blanco


### Ejercicio 6

Crea una clase _Animal_ con un atributo _name_ y un método _speak_ que imprima por pantalla el mensaje "Yo soy {nombre}".

Además crea una clase _Dog_ que herede de la clase _Animal_ y que tenga un atributo _raza_ y un método _speak_ que imprima por pantalla el mensaje "Yo soy {nombre} y soy un perro de raza {raza}, y ladro".

Crea un objeto de la clase _Dog_ y comprueba que el método _speak_ funciona correctamente.

In [3]:
class Animal:
    def __init__(self, name):
        self.name = name
    
    def nombre(self):
        print(f"Yo soy {self.nombre}")

class Dog(Animal):
    def __init__(self, name, raza):
        Animal.__init__(self, name)
        self.raza = raza
    
    def speak(self):
        print(f"Yo soy {self.name} y soy un perro de raza {self.raza}, y ladro")

mastin = Dog("Rufo", "mastín")
mastin.speak()

Yo soy Rufo y soy un perro de raza mastín, y ladro


### Ejercicio 7

Basado en el ejercicio anterior, en la el constructor  de la clase _Dog_ utiliza la función _super()_ para llamar al constructor de la clase _Animal_ y asignar el atributo _name_.

Crea un objeto de la clase _Dog_ y comprueba que el método _speak_ funciona correctamente.

In [48]:
class Animal:
    def __init__(self, name):
        self.name = name
    
    def nombre(self):
        print(f"Yo soy {self.nombre}")

class Dog(Animal):
    def __init__(self, name, raza):
        super().__init__(name)
        self.raza = raza
    
    def speak(self):
        print(f"Yo soy {self.name} y soy un perro de raza {self.raza}, y ladro")

mastin = Dog("Rufo", "mastín")
mastin.speak()

Yo soy Rufo y soy un perro de raza mastín, y ladro


### Ejercicio 8

Crear una clase _Circle_ con un atributo _radius_, este atributo debe ser privado.<br>
Crea una función getter y setter para el atributo _radius_, que permita obtener y asignar un valor al atributo.<br>
Es necesario comprobar que el valor asignado al atributo es un número positivo, en caso contrario no asigna el valor.

Crea un objeto de la clase _Circle_ y comprueba que el método _setter_ funciona correctamente, y que no permite asignar un valor negativo.

In [54]:
class Circle:
    def __init__(self, radius=0):
        self._radius = radius

    
    def get_radio(self):
        return self._radius
    
    
    def set_radio(self, valor):
        if valor >= 0:
            self._radius = valor
            return self._radius

circulo = Circle()

print(circulo.set_radio(-4))

        


None


### Ejercicio 9

Basado en el ejercicio anterior, modifica la clase _Circle_, cambiando las funciones getter y setter par la propiedad _radius_ con la forma de _property_.

In [33]:
class Circle:
    def __init__(self, radius = 0):
        self._radius = radius

    @property
    def radio(self):
        return self._radius
    
    @radio.setter
    def radio(self, valor):
        if valor >= 0:
            self._radius = valor

circulo = Circle()
circulo.radio = -2
print(circulo.radio)

0


### Ejercicio 10

Crea una clase _Staff_ con un atributo _role_, _department_ y _salary_.<br>
La clase _Staff_ debe tener un método _info_ que muestre la información del empleado.<br>

Crea una clase _Teacher_ que herede de la clase _Staff_ y que tenga un atributo _name_ y _age_.<br>
Para el constructor de la clase _Teacher_ utiliza la función _super()_ para llamar al constructor de la clase _Staff_ y asignar los atributos _role_, _department_ y _salary_.<br>
Override el método _info_ para que muestre la información del profesor, incluyendo el nombre y la edad.
Crea otro método _infoSuper_ que utilize la sintaxis de super para llamar al método _info_ de la clase padre.

Crea un objeto de la clase _Teacher_ y comprueba que el método _info_ funciona correctamente.


In [57]:
class Staff:
    def __init__(self, role, department, salary):
        self.role = role
        self.department = department
        self.salary = salary
    
    def info(self):
        print(self.role, self.department, self.salary)

class Teacher(Staff):
    def __init__(self, role, department, salary, name, age):
        super().__init__(role, department, salary)
        self.name = name
        self.age = age
    
    def info(self):
        print(self.role, self.department, self.salary, self.name, self.age)
    
    def infoSuper(self):
        super().info(self)

profesor = Teacher("Profesor", "Fisica", "1300€", "Paco", "59")
profesor.info()

Profesor Fisica 1300€ Paco 59


#### Ejercicio 11

Crea una clase `BanckAccount` con atributos balance y account_number, estos métodos deben ser privados.<br>
Implementa métodos para obtener el número de cuenta, obtener el balance, depositar fondos y retirar fondos, utilizando getters y setters.

Validaciones:
- El número de cuenta debe ser de solo lectura.
- El balance no puede ser negativo.
- El método para retirar fondos no puede retirar más fondos de los que hay en la cuenta.
- El método para retirar fondos debe devolver el dinero retirado, pero si no hay suficiente dinero en la cuenta debe devolver el dinero que hay en la cuenta.

Crea un objeto de la clase `BanckAccount` y comprueba que los métodos funcionan correctamente, con cada uno de los casos de validación.

In [3]:
class BanckAccount:
    def __init__(self, balance = 0, account_number = "00000000"):
        self._balance = balance
        self._account_number = account_number
        
    def get_numeroCuenta(self):
        return self._account_number

    def get_balance(self):
        return self._balance

    # Metodos    
    def Retirar(self, cantidad):
        if cantidad > 0:
            if cantidad <= self._balance:
                self._balance -= cantidad
                print(f"Saldo disponible en la cuenta {self._balance}")
            else:
                print("No puede retirar una catidad superior al total de la cuenta")
                print(self._balance)
        else:
            print("No puede retirar una cantidad menor de 0")

    def Ingresar(self, cantidad):
        if cantidad > 0:
            self._balance += cantidad
            return f"Saldo disponible en la cuenta {self._balance}"
        else:
            return "El ingreso debe ser mayor que cero"

cc = BanckAccount(10000, "235624563")

print(f"Numero de cuenta {cc.get_numeroCuenta()}")
print(f"Saldo {cc.get_balance()}")

cc.Ingresar(5000)
print(cc.get_balance())
cc.Retirar(395)
cc.Retirar(15555555)
cc.Ingresar(-26)




    


Numero de cuenta 235624563
Saldo 10000
15000
Saldo disponible en la cuenta 14605
No puede retirar una catidad superior al total de la cuenta
14605


'El ingreso debe ser mayor que cero'

#### Ejercicio 12

Basado en el ejercicio anterior, cambia los getters y setters utilizando la sintaxis de property (decoradores).

Comprueba que todo funciona correctamente, al igual que en el ejercicio anterior.

In [ ]:
class BanckAccount:
    def __init__(self, balance = 0, account_number = "00000000"):
        self._balance = balance
        self._account_number = account_number
        
    @property
    def numeroCuenta(self):
        return self._account_number
    
    @property
    def balance(self):
        return self._balance

    # Metodos    
    def Retirar(self, cantidad):
        if cantidad > 0:
            if cantidad <= self._balance:
                self._balance -= cantidad
                print(f"Saldo disponible en la cuenta {self._balance}")
            else:
                print("No puede retirar una catidad superior al total de la cuenta")
                print(self._balance)
        else:
            print("No puede retirar una cantidad menor de 0")

    def Ingresar(self, cantidad):
        if cantidad > 0:
            self._balance += cantidad
            return f"Saldo disponible en la cuenta {self._balance}"
        else:
            return "El ingreso debe ser mayor que cero"

cc = BanckAccount(10000, "235624563")

print(f"Numero de cuenta {cc.numeroCuenta}")
print(f"Saldo {cc.balance}")

cc.Ingresar(5000)
print(cc.balance)
cc.Retirar(395)
cc.Retirar(15555555)
cc.Ingresar(-26)

#### Ejercicio 13

Crea una clase `Person` que represente la información de una persona, incluyendo su nombre, edad y email.<br>
Sin embargo para mantener la privacidad se necesita usar getters y setters para acceder y modificar estos atributos.<br>

Reglas:
- El nombre y el email no pueden ser modificados.
- La edad debe ser un número positivo entre 0 y 120.

Crea un objeto de la clase `Person` y comprueba que los métodos funcionan correctamente, con cada uno de los casos de validación.

In [17]:
class Person:
    def __init__(self, nombre, edad, email):
        self._nombre = nombre
        self._edad = edad
        self._email = email

    def get_edad(self):
        return self._edad
    
    def set_edad(self, valor):
        if valor > 0:
            self._edad = valor
        else:
            print("Edad debe ser mayor que Cero")

    def get_nombre(self):
        return self._nombre
    
    def get_email(self):
        return self._email
    
Manuel = Person("Manuel Lopez", 48, "mlopez@gmail.com")

print(Manuel.get_edad())
Manuel.set_edad(49)
print(Manuel.get_nombre(), Manuel.get_edad(), Manuel.get_email())
Manuel.set_edad(-2)



48
Manuel Lopez 49 mlopez@gmail.com
Edad debe ser mayor que Cero
